### ISSUE TITLE
[FR] Option for grep to return names and maybe a kvp (name, uri)

### ISSUE THREAD

**Issue by ajboni** [SpotlightKid/jack-audio-tools#1]
Hi!! Thanks for your work here!
I'm writing an app which use this library to get plugins info, my idea is to make an array of objects with name and uri.

```js
[
  {
    name: "ZamVerb",
    uri: "urn:zamaudio:ZamVerb",
  },
  {
    name: "ZamTube",
    uri: "urn:zamaudio:ZamTube"
  }
]
```
I can get the info with `lv2ls` but it would be nice to have it all in the  same package.
The use case is to load only names and uris, and only request, call `plugin_info(uri)` to avoid building the database with `plugins_info()` which may take a while.

**Comment by SpotlightKid** [comment]
You mean like this? #2

**Comment by SpotlightKid** [comment]
This still only searches through the URI, though. Otherwise the full TTL of every plugin would have to be loaded, which slows the script down too much, for it to be useful anymore for its original purpose, which is quickly finding the URI of a plugin.

**Comment by ajboni** [comment]
that was fast!! and that's perfect! 
could it be possible to load the category entries without loading the ttl like you did with the name? 
(if yes I would research the topic and do a PR if you agree ) 
Im loading each plugin ttl to make a 'category' database, but  it is not quick, as expected...

**Comment by SpotlightKid** [comment]
> could it be possible to load the category entries without loading the ttl like you did with the name?

For most plugins lilv has to load the plugin's ttl for everything beside the plugin URI, because `manifest.ttl` will contain nothing more than a `rdfs:seeAlso` to the full ttl file.

Here's a quick script to list all plugins with their categories:

```py
#!/usr/bin/env python
"""List all LV2 plugins with their categories."""

import json
import sys

import lilv


def main():
    world = lilv.World()
    world.load_all()

    results = []
    for plugin in world.get_all_plugins():
        uri = str(plugin.get_uri())
        world.load_resource(uri)
        name = plugin.get_name()
        categories = []
        for cat in plugin.get_value(world.ns.rdf.type):
            cat = str(cat)
            if not cat.endswith('#Plugin'):
                categories.append(str(cat).split('#', 1)[-1])

        results.append({
            'name': str(name) if name is not None else None,
            'uri': uri,
            'categories': categories
        })

    json.dump(results, sys.stdout, indent=2)
    print("Found {} LV2 plugins.".format(len(results)), file=sys.stderr)


if __name__ == '__main__':
    sys.exit(main() or 0)
```

On my system this takes about 3 seconds to complete:

```console
$ time python lv2/list_plugin_categories.py >/dev/null
Found 1130 LV2 plugins.

real	0m2,877s
user	0m2,778s
sys	0m0,087s
```

**Comment by ajboni** [comment]
That's definitely much faster than using `lv2.plugin_info`! 
It is correct to assume that what 'costs' is to do the query and not load the TTL file then?

```bash 
$ time python3 -m jackaudiotools.lv2.plugin_info >/dev/null

Executed in   12,40 secs   fish           external 
   usr time   12,29 secs  329,00 micros   12,29 secs 
   sys time    0,11 secs   97,00 micros    0,11 secs 
```
I can definitely afford 3~5 seconds on boot if that means keeping the plugin database up-to date.
Are you interested on adding this in jackaudiotools?

---
offtopic (maybe for another FR):
what about a similar function that we could pass an array of fields to query as arguments? We could build the JSON with fields that we need at the moment, (assuming that loading the TTL is cheaper than querying)

**Comment by SpotlightKid** [comment]
> It is correct to assume that what 'costs' is to do the query and not load the TTL file then?

No. `plugin_info.py` gathers much more information on each plugin, including examining all ports, properties and looking for presets. Especially the last action can potentially lead to loading lots more  TTL files. I haven't really profiled it but I suspect that's where much of the additional time is spent.

> what about a similar function that we could pass an array of fields to query as arguments?

My idea about this project is that is really only a collection of utilities and code examples of how to use JACK-Client and lilv and provide some helpers for audio users working on the console. It isn't really meant as a library to be used by other programs. If you find code that your program can use, fine, just copy it into your project (with proper attribution). Or use the existing code as inspiration to expand upon.

**Comment by SpotlightKid** [comment]
(cont.) Though it might be worthwhile to put certain functions (like listing all plugins with names and or catgories) into a a module that can be used by several scripts. Sort of a convenience wrapper for lilv, which could grow over time.

**Comment by ajboni** [comment]
> My idea about this project is that is really only a collection of utilities and code examples of how to use JACK-Client and lilv and provide some helpers for audio users working on the console. It isn't really meant as a library to be used by other programs. If you find code that your program can use, fine, just copy it into your project (with proper attribution). Or use the existing code as inspiration to expand upon.

Understood!

> (cont.) Though it might be worthwhile to put certain functions (like listing all plugins with names and or catgories) into a a module that can be used by several scripts. Sort of a convenience wrapper for lilv, which could grow over time.

That's how I found the project actually, looking for a 'convenience wrapper for lilv'.

Anyway, thanks again for your work and replies! I will follow your advice and try to put together some useful helper functions, using this project as base/reference. Thanks!

**Comment by SpotlightKid** [comment]
> Are you interested on adding this in jackaudiotools?

I just merged #2 where I added this as an extra option to `lv2-grep`.

**Comment by ajboni** [comment]
> > Are you interested on adding this in jackaudiotools?
> 
> I just merged #2 where I added this as an extra option to `lv2-grep`.

Thank you, this is really useful!
